### ISSUE TITLE
numerical unstable log(softmax(·))

### ISSUE THREAD

**Issue by AndreasMadsen** [Theano/Theano#2781]
This example results in an `inf` loss:

```python
x = T.cast(T.arange(1, 10), 'float32') * 13
y = T.nnet.softmax(x)
t = T.unbroadcast(T.zeros(1, dtype='int64'), 0)
L = T.mean(T.nnet.categorical_crossentropy(y, t))

loss = theano.function(
    inputs=[], outputs=[L]
)
print(loss())
```

If done as sugested in https://groups.google.com/forum/#!topic/theano-dev/Vg08s2UegHU, that is use something equivalent to:

```python
def log_softmax(x):
   x_diff = x - T.max(x)
   return x_diff - T.log(T.sum(T.exp(x_diff)))
```

then one gets `-104.0` which is somewhat close to the correct `-101.0`.

**Comment by lamblin** [comment]
This is due to premature constant folding when optimizing the graph, since the value of your input is a constant. If you use a shared variable or an explicit input, it does not happen:
```python

x = theano.shared((np.arange(1, 10) * 13).astype('float32'))
y = T.nnet.softmax(x)
t = T.unbroadcast(T.zeros(1, dtype='int64'), 0)
L = T.mean(T.nnet.categorical_crossentropy(y, t))

loss = theano.function(
    inputs=[], outputs=[L]
)
print(loss())
```
gives `[array(104.0, dtype=float32)]`.

**Comment by AndreasMadsen** [comment]
Hmm, okay. In the original code the input isn't constant and I still have a numerical stability issue. I guess this is because the `softmax` is an `.scan` and the `categorical_crossentropy` is outside the scan, like this:

```python
x = theano.shared((np.arange(1, 10) * 3).astype('float32'))
(y, _) = theano.scan(lambda t_i: T.nnet.softmax(x * t_i), T.arange(0, 6))
y = y.reshape((y.shape[0], y.shape[1] * y.shape[2]))

t = T.zeros(10, dtype='int64')
L = T.mean(T.nnet.categorical_crossentropy(y, t))

loss = theano.function(
    inputs=[], outputs=[L]
)
print(loss())
```

this gives `[array(inf, dtype=float32)]`.

**Comment by lamblin** [comment]
See #2944.
